File: src/frontend/cte/lib/app/data/model/cte_aereo_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:cte/app/infra/infra_imports.dart';
import 'package:intl/intl.dart';
import 'package:cte/app/data/domain/domain_imports.dart';

class CteAereoModel {
	int? id;
	int? idCteCabecalho;
	int? numeroMinuta;
	int? numeroConhecimento;
	DateTime? dataPrevistaEntrega;
	String? idEmissor;
	String? idInternaTomador;
	String? tarifaClasse;
	String? tarifaCodigo;
	double? tarifaValor;
	String? cargaDimensao;
	String? cargaInformacaoManuseio;
	String? cargaEspecial;

	CteAereoModel({
		this.id,
		this.idCteCabecalho,
		this.numeroMinuta,
		this.numeroConhecimento,
		this.dataPrevistaEntrega,
		this.idEmissor,
		this.idInternaTomador,
		this.tarifaClasse,
		this.tarifaCodigo,
		this.tarifaValor,
		this.cargaDimensao,
		this.cargaInformacaoManuseio,
		this.cargaEspecial,
	});

	static List<String> dbColumns = <String>[
		'id',
		'numero_minuta',
		'numero_conhecimento',
		'data_prevista_entrega',
		'id_emissor',
		'id_interna_tomador',
		'tarifa_classe',
		'tarifa_codigo',
		'tarifa_valor',
		'carga_dimensao',
		'carga_informacao_manuseio',
		'carga_especial',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Numero Minuta',
		'Numero Conhecimento',
		'Data Prevista Entrega',
		'Id Emissor',
		'Id Interna Tomador',
		'Tarifa Classe',
		'Tarifa Codigo',
		'Tarifa Valor',
		'Carga Dimensao',
		'Carga Informacao Manuseio',
		'Carga Especial',
	];

	CteAereoModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idCteCabecalho = jsonData['idCteCabecalho'];
		numeroMinuta = jsonData['numeroMinuta'];
		numeroConhecimento = jsonData['numeroConhecimento'];
		dataPrevistaEntrega = jsonData['dataPrevistaEntrega'] != null ? DateTime.tryParse(jsonData['dataPrevistaEntrega']) : null;
		idEmissor = jsonData['idEmissor'];
		idInternaTomador = jsonData['idInternaTomador'];
		tarifaClasse = CteAereoDomain.getTarifaClasse(jsonData['tarifaClasse']);
		tarifaCodigo = jsonData['tarifaCodigo'];
		tarifaValor = jsonData['tarifaValor']?.toDouble();
		cargaDimensao = jsonData['cargaDimensao'];
		cargaInformacaoManuseio = CteAereoDomain.getCargaInformacaoManuseio(jsonData['cargaInformacaoManuseio']);
		cargaEspecial = CteAereoDomain.getCargaEspecial(jsonData['cargaEspecial']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idCteCabecalho'] = idCteCabecalho != 0 ? idCteCabecalho : null;
		jsonData['numeroMinuta'] = numeroMinuta;
		jsonData['numeroConhecimento'] = numeroConhecimento;
		jsonData['dataPrevistaEntrega'] = dataPrevistaEntrega != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataPrevistaEntrega!) : null;
		jsonData['idEmissor'] = idEmissor;
		jsonData['idInternaTomador'] = idInternaTomador;
		jsonData['tarifaClasse'] = CteAereoDomain.setTarifaClasse(tarifaClasse);
		jsonData['tarifaCodigo'] = tarifaCodigo;
		jsonData['tarifaValor'] = tarifaValor;
		jsonData['cargaDimensao'] = cargaDimensao;
		jsonData['cargaInformacaoManuseio'] = CteAereoDomain.setCargaInformacaoManuseio(cargaInformacaoManuseio);
		jsonData['cargaEspecial'] = CteAereoDomain.setCargaEspecial(cargaEspecial);
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idCteCabecalho = plutoRow.cells['idCteCabecalho']?.value;
		numeroMinuta = plutoRow.cells['numeroMinuta']?.value;
		numeroConhecimento = plutoRow.cells['numeroConhecimento']?.value;
		dataPrevistaEntrega = Util.stringToDate(plutoRow.cells['dataPrevistaEntrega']?.value);
		idEmissor = plutoRow.cells['idEmissor']?.value;
		idInternaTomador = plutoRow.cells['idInternaTomador']?.value;
		tarifaClasse = plutoRow.cells['tarifaClasse']?.value != '' ? plutoRow.cells['tarifaClasse']?.value : 'AAA';
		tarifaCodigo = plutoRow.cells['tarifaCodigo']?.value;
		tarifaValor = plutoRow.cells['tarifaValor']?.value?.toDouble();
		cargaDimensao = plutoRow.cells['cargaDimensao']?.value;
		cargaInformacaoManuseio = plutoRow.cells['cargaInformacaoManuseio']?.value != '' ? plutoRow.cells['cargaInformacaoManuseio']?.value : 'AAA';
		cargaEspecial = plutoRow.cells['cargaEspecial']?.value != '' ? plutoRow.cells['cargaEspecial']?.value : 'AAA';
	}	

	CteAereoModel clone() {
		return CteAereoModel(
			id: id,
			idCteCabecalho: idCteCabecalho,
			numeroMinuta: numeroMinuta,
			numeroConhecimento: numeroConhecimento,
			dataPrevistaEntrega: dataPrevistaEntrega,
			idEmissor: idEmissor,
			idInternaTomador: idInternaTomador,
			tarifaClasse: tarifaClasse,
			tarifaCodigo: tarifaCodigo,
			tarifaValor: tarifaValor,
			cargaDimensao: cargaDimensao,
			cargaInformacaoManuseio: cargaInformacaoManuseio,
			cargaEspecial: cargaEspecial,
		);			
	}

	
}

File: backend/cte/src/model/cte_aereo_model.py
from src import db


class CteAereoModel(db.Model):
    __tablename__ = 'cte_aereo'

    id = db.Column(db.Integer, primary_key=True)
    numero_minuta = db.Column(db.Integer)
    numero_conhecimento = db.Column(db.Integer)
    data_prevista_entrega = db.Column(db.DateTime)
    id_emissor = db.Column(db.String(20))
    id_interna_tomador = db.Column(db.String(14))
    tarifa_classe = db.Column(db.String(1))
    tarifa_codigo = db.Column(db.String(4))
    tarifa_valor = db.Column(db.Float)
    carga_dimensao = db.Column(db.String(14))
    carga_informacao_manuseio = db.Column(db.String(1))
    carga_especial = db.Column(db.String(3))
    id_cte_cabecalho = db.Column(db.Integer, db.ForeignKey('cte_cabecalho.id'))


    def mapping(self, data):
        self.id = data.get('id')
        self.id_cte_cabecalho = data.get('idCteCabecalho')
        self.numero_minuta = data.get('numeroMinuta')
        self.numero_conhecimento = data.get('numeroConhecimento')
        self.data_prevista_entrega = data.get('dataPrevistaEntrega')
        self.id_emissor = data.get('idEmissor')
        self.id_interna_tomador = data.get('idInternaTomador')
        self.tarifa_classe = data.get('tarifaClasse')
        self.tarifa_codigo = data.get('tarifaCodigo')
        self.tarifa_valor = data.get('tarifaValor')
        self.carga_dimensao = data.get('cargaDimensao')
        self.carga_informacao_manuseio = data.get('cargaInformacaoManuseio')
        self.carga_especial = data.get('cargaEspecial')

    def serialize(self):
        return {
            'id': self.id,
            'idCteCabecalho': self.id_cte_cabecalho,
            'numeroMinuta': self.numero_minuta,
            'numeroConhecimento': self.numero_conhecimento,
            'dataPrevistaEntrega': self.data_prevista_entrega.isoformat(),
            'idEmissor': self.id_emissor,
            'idInternaTomador': self.id_interna_tomador,
            'tarifaClasse': self.tarifa_classe,
            'tarifaCodigo': self.tarifa_codigo,
            'tarifaValor': self.tarifa_valor,
            'cargaDimensao': self.carga_dimensao,
            'cargaInformacaoManuseio': self.carga_informacao_manuseio,
            'cargaEspecial': self.carga_especial,
        }